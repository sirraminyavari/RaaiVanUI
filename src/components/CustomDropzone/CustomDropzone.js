/**
 * Renders a file upload conponent.
 */
import { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { useDropzone } from 'react-dropzone';
import * as Styled from './CustomDropzone.styles';
import UploadFileIcon from 'components/Icons/UploadFileIcon/UploadFile';
import APIHandler from 'apiHelper/APIHandler';
import axios from 'axios';
import ProgressBar from 'components/ProgressBar/ProgressBar';

/**
 * @typedef PropType
 * @property {string} accept -The accepted file format.
 * @property {number} maxFiles -The maximum number of files that can be uploaded.
 * @property {number} maxEachSize -The maximum size of each file.
 * @property {number} maxTotalSize -The maximum size of all files.
 * @property {function} onError -A callback function that will fire on file upload exception.
 * @property {Object} containerProps -The props passed to dropzone container.
 * @property {Object} inputProps -The props passed to input.
 * @property {string} [nodeId] -Node id.
 * @property {boolean} disabled -A flag that will disable dropzone area.
 * @property {string[]} exceptions -All formats that are not allowed to be uploaded.
 * {exceptions} prop has priority over {accept} prop.
 * If a format is defined forbidden in exceptions, component will throw error even if it is accepted.
 */

/**
 *  @description Renders a file upload component.
 * @component
 * @param {PropType} props
 */
const CustomDropzone = (props) => {
  const {
    accept,
    maxFiles,
    maxEachSize,
    maxTotalSize,
    onError,
    containerProps,
    inputProps,
    nodeId,
    disabled,
    exceptions,
  } = props;
  const [files, setFiles] = useState([]);
  const [errors, setErrors] = useState([]);
  const [bar, setBar] = useState({});
  const apiHandler = new APIHandler('DocsAPI', 'GetUploadLink');

  //! Converts bytes to mega bytes
  const sizeInMB = (sizeInBytes) =>
    Number((sizeInBytes / (1024 * 1024)).toFixed(2));

  //! Custom validator
  const customValidator = (file) => {
    const fileFormat = file.name.split('.').pop();
    const fileSize = file.size;
    const fileName = file.name;

    //! Compare file size against maximum allowed size for each file.
    if (sizeInMB(fileSize) > maxEachSize) {
      return {
        code: 'file-too-large',
        message: `حجم فایل ${fileName} بزرگتر از حد مجاز (${maxEachSize} مگابایت)است.`,
      };
    }

    //! See if file format is allowed or not.
    if (exceptions && exceptions.includes(fileFormat)) {
      return {
        code: 'not-allowed-format',
        message: `فایل با فرمت ${fileFormat} مجاز نمی باشد.`,
      };
    }

    return null; //! file is clean and there is no error.
  };

  //! Sets toastify alert for each error.
  useEffect(() => {
    if (errors.length) {
      errors.forEach((err) => {
        window.alert(err.message);
      });
    }
  }, [errors]);

  //! Fires on drop event.
  const onDrop = useCallback((acceptedFiles) => {
    const totalSize = acceptedFiles
      .map((file) => file.size)
      .reduce((sum, size) => sum + sizeInMB(size), 0);

    if (totalSize > maxTotalSize) {
      //! Creates maximum total size error.
      const totalSizeError = {
        code: 'total-size-exception',
        message: `حجم کل فایل ها بیشتر از حد مجاز (${maxTotalSize} مگابایت) است.`,
      };
      // setFiles((files) => {
      //   files.length = 0;
      //   return files;
      // });
      setErrors((c) => c.concat(totalSizeError));
    } else {
      //! Sets accepted files and assign new properties.
      setFiles(
        acceptedFiles.map((file) =>
          Object.assign(file, {
            preview: URL.createObjectURL(file),
            uploadPercentage: 0,
          })
        )
      );
      console.log(acceptedFiles);
      //! Do something with the files

      //! Uploads accepted files asynchronous.
      acceptedFiles.reduce(async (previousPromise, accepted) => {
        await previousPromise;

        return new Promise((resolve, reject) => {
          //! Get upload link.
          apiHandler.url(
            {
              OwnerID: nodeId,
              OwnerType: 'Node',
            },
            (response) => {
              let uploadURL = response.slice(5);
              //! Prepare upload data
              let formData = new FormData();
              formData.append('file', accepted);

              //! Keep track of upload progress.
              let options = {
                onUploadProgress: (progressEvent) => {
                  const { loaded, total } = progressEvent;
                  let percentage = Math.floor((loaded * 100) / total);

                  //! Sets current progress bar values.
                  setBar({
                    name: accepted.name,
                    uploadPercentage: percentage,
                    description: `${sizeInMB(loaded)}MB of ${sizeInMB(
                      total
                    )}MB`,
                  });

                  //! Executes if upload progress is over.
                  if (percentage === 100) {
                    //! Sets upload percentage property of just finished file to 100.
                    setFiles((c) =>
                      c.map((file) => {
                        if (file.name === accepted.name) {
                          file.uploadPercentage = 100;
                          return file;
                        }
                        return file;
                      })
                    );
                    //! Empty finished progress bar.
                    setBar({});
                    resolve();
                  }
                },
              };
              //! Post file to server.
              // axios.post(uploadURL, formData, options);
            }
          );
        });
      }, Promise.resolve());
    }
  }, []);

  const {
    getRootProps,
    getInputProps,
    isDragActive,
    fileRejections,
    acceptedFiles,
    // open,
  } = useDropzone({
    onDrop,
    accept,
    maxFiles,
    disabled: !!disabled,
    validator: customValidator,
    // noKeyboard: true,
    // noClick: true,
  });

  //! Handle errors thrown from dropzone and validator.
  const handlErrors = () => {
    //! Maximum file number error.
    if (fileRejections.length) {
      // console.log(fileRejections);
      if (
        fileRejections.some((file) => file.errors[0].code === 'too-many-files')
      ) {
        setErrors((c) =>
          c.concat({
            code: fileRejections[0].errors[0].code,
            message: `تعداد فایل ها بیشتر از حد مجاز (${maxFiles} عدد) است.`,
          })
        );
      }

      //! Maximum file size error.
      if (
        fileRejections.some((file) => file.errors[0].code === 'file-too-large')
      ) {
        const fileSizeError = fileRejections
          .filter((file) => file.errors[0].code !== 'too-many-files')
          .map((file) => {
            return {
              code: file.errors[0].code,
              message: file.errors[0].message,
              fileName: file.file.name,
              fileSize: {
                sizeInBytes: file.file.size,
                sizeInMB: sizeInMB(file.file.size),
                sizeReadable: sizeInMB(file.file.size) + 'MB',
              },
            };
          });
        setErrors((c) => c.concat(fileSizeError));
      }
    }

    //! Format exceptions error.
    if (
      fileRejections.some(
        (file) => file.errors[0].code === 'not-allowed-format'
      )
    ) {
      setErrors((c) =>
        c.concat({
          code: fileRejections[0].errors[0].code,
          message: fileRejections[0].errors[0].message,
        })
      );
    }
  };

  //! Prepares thumbnail images.
  const thumbs = files
    .filter((file) => file.name.endsWith('.jpg') || file.name.endsWith('.png'))
    .map((file) => (
      <Styled.Thumb key={file.name}>
        <Styled.ThumbInner>
          <Styled.ThumbImage src={file.preview} alt="thumb-dropzone" />
        </Styled.ThumbInner>
      </Styled.Thumb>
    ));

  useEffect(() => {
    handlErrors();
    //! Clearn up
    return () => {
      //! Make sure to revoke the data uris to avoid memory leaks
      files.forEach((file) => URL.revokeObjectURL(file.preview));
      setErrors([]);
    };
  }, [files]);

  useEffect(() => {
    if (errors.length) {
      onError(errors);
    }
  }, [errors]);

  return (
    <>
      <Styled.DropzoneContainer
        {...getRootProps({ ...containerProps, refKey: 'innerRef' })}>
        <Styled.UploadIconWrapper>
          <UploadFileIcon size={30} color="#fff" />
        </Styled.UploadIconWrapper>
        <Styled.InputWrapper>
          <input {...getInputProps(inputProps)} />
          {isDragActive ? (
            <p>فایل های انتخابی را در این باکس رها کنید...</p>
          ) : (
            <p>
              برای انتخاب فایل آنها را بگیرید و درون باکس رها و یا کلیک کنید...
            </p>
          )}
        </Styled.InputWrapper>
      </Styled.DropzoneContainer>
      <Styled.ThumbsContainer>{thumbs}</Styled.ThumbsContainer>
      {errors.map((error, index) => {
        return <p style={{ color: 'red' }}>{error.message}</p>;
      })}
      {files.map((file, index, self) => {
        if (file.name === bar.name) {
          return (
            <ProgressBar
              progress={bar.uploadPercentage}
              label={`${bar.name}(${bar.description})`}
            />
          );
        }
        return (
          <ProgressBar progress={file.uploadPercentage} label={file.name} />
        );
      })}
    </>
  );
};

CustomDropzone.propTypes = {
  accept: PropTypes.string,
  maxFiles: PropTypes.number,
  maxEachSize: PropTypes.number,
  maxTotalSize: PropTypes.number,
  onError: PropTypes.func,
  containerProps: PropTypes.object,
  inputProps: PropTypes.object,
  nodeId: PropTypes.string,
  disabled: PropTypes.bool,
  exceptions: PropTypes.array,
};

CustomDropzone.defaultProps = {
  nodeId: '',
};

CustomDropzone.displayName = 'CustomDropzoneComponent';

export default CustomDropzone;
