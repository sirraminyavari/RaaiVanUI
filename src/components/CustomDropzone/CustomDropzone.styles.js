import styled from 'styled-components';

export const DropzoneContainer = styled.div`
  border: 0.15rem dashed #2b388f;
  margin: 0 1rem;
  padding: 0.5rem 1rem;
  border-radius: 0.3rem;
  display: flex;
  justify-content: end;
  align-items: center;
`;

export const UploadIconWrapper = styled.div`
  border: 2px solid transparent;
  display: inline-block;
  border-radius: 50%;
  line-height: 0.8rem;
  padding: 0.2rem 0.3rem;
  background-color: #2b388f;
`;

export const InputWrapper = styled.div`
  font-size: 1rem;
  margin: 0 1rem;
`;

export const ThumbsContainer = styled.aside`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  margin-top: 1rem;
`;

export const Thumb = styled.div`
  display: inline-flex;
  border-radius: 0.3rem;
  border: 0.2rem solid #999;
  margin-bottom: 0.5rem;
  margin-right: 0.5rem;
  width: 8rem;
  height: 8rem;
  padding: 0.4rem;
  box-sizing: border-box;
`;

export const ThumbInner = styled.div`
  display: flex;
  min-width: 0;
  overflow: hidden;
`;

export const ThumbImage = styled.img`
  display: block;
  width: auto;
  height: 100%;
  border-radius: 0.3rem;
`;
