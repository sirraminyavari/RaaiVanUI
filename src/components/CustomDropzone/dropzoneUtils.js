export const errorTypes = {
  LARGE_FILE_ERROR: 'file-too-large',
  FORMAT_ERROR: 'not-allowed-format',
  TOTAL_SIZE_ERROR: 'total-size-exception',
  TOO_MANY_FILES_ERROR: 'too-many-files',
};
