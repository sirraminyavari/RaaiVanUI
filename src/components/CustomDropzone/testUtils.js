import { act, fireEvent, waitFor } from '@testing-library/react';

//! Provides mock file for tests.
export const createFile = (name, size, type) => {
  const file = new File([], name, { type });

  Object.defineProperty(file, 'size', {
    get() {
      return size;
    },
  });

  return file;
};

//! Using fireEvent.* doesn't work for our use case,
//! we cannot set the event props
export const dispatchEvt = (node, type, data) => {
  const event = new Event(type, { bubbles: true });
  if (data) {
    Object.assign(event, data);
  }
  fireEvent(node, event);
};

//! Dispatches an event on drag enter.
export const fireDragEnter = (node, data) => {
  dispatchEvt(node, 'dragenter', data);
};

//! Dispatches an event on drag over.
export const fireDragOver = (node, data) => {
  dispatchEvt(node, 'dragover', data);
};

//! Dispatches an event on drag leave.
export const fireDragLeave = (node, data) => {
  dispatchEvt(node, 'dragleave', data);
};

//! Dispatches an event on drop.
export const fireDrop = (node, data) => {
  dispatchEvt(node, 'drop', data);
};

export const flushPromises = async (rerender, ui) => {
  await act(() => waitFor(() => rerender(ui)));
};

//! The DataTransfer object is used to hold the data that is being dragged during a drag and drop operation.
//! It may hold one or more data items, each of one or more data types.
//! See: https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer
export const createDataTransferWithFiles = (files = []) => {
  return {
    dataTransfer: {
      files,
      items: files.map((file) => ({
        kind: 'file',
        size: file.size,
        type: file.type,
        getAsFile: () => file,
      })),
      types: ['Files'],
    },
  };
};
