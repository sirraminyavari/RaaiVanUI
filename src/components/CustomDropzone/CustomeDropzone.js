/**
 * Renders a file upload conponent.
 */
import { useCallback, useEffect, useState } from 'react';
import { useDropzone } from 'react-dropzone';
import * as Styled from './CustomDropzone.styles';
import UploadFileIcon from 'components/Icons/UploadFileIcon/UploadFile';

const CustomeDropzone = (props) => {
  const { accept } = props;
  const [files, setFiles] = useState([]);

  const onDrop = useCallback((acceptedFiles) => {
    setFiles(
      acceptedFiles.map((file) =>
        Object.assign(file, {
          preview: URL.createObjectURL(file),
        })
      )
    );
    //! Do something with the files
    acceptedFiles.forEach((file) => {
      const reader = new FileReader();

      reader.readAsDataURL(file);

      reader.onabort = () => console.log('file reading was aborted');
      reader.onerror = () => console.log('file reading has failed');
      reader.onload = () => {
        //! Do whatever you want with the file contents
        const fileURL = reader.result;
        console.log(fileURL);
      };
    });
  }, []);

  const { getRootProps, getInputProps, isDragActive, open } = useDropzone({
    onDrop,
    accept,
    // noKeyboard: true,
    // noClick: true,
  });

  const thumbs = files.map((file) => (
    <Styled.Thumb key={file.name}>
      <Styled.ThumbInner>
        <Styled.ThumbImage src={file.preview} alt="thumb-dropzone" />
      </Styled.ThumbInner>
    </Styled.Thumb>
  ));

  useEffect(
    () => () => {
      // Make sure to revoke the data uris to avoid memory leaks
      files.forEach((file) => URL.revokeObjectURL(file.preview));
    },
    [files]
  );

  return (
    <>
      <Styled.DropzoneContainer {...getRootProps({ refKey: 'innerRef' })}>
        <Styled.UploadIconWrapper>
          <UploadFileIcon size={30} color="#fff" />
        </Styled.UploadIconWrapper>
        <Styled.InputWrapper>
          <input {...getInputProps()} />
          {isDragActive ? (
            <p>فایل های انتخابی را در این باکس رها کنید...</p>
          ) : (
            <p>
              برای انتخاب فایل آنها را بگیرید و اینجا رها کنید و یا درون باکس
              کلیک کنید...
            </p>
          )}
        </Styled.InputWrapper>
      </Styled.DropzoneContainer>
      <Styled.ThumbsContainer>{thumbs}</Styled.ThumbsContainer>
    </>
  );
};

export default CustomeDropzone;
