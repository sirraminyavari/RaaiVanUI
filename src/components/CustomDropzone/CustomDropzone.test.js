import ReactDOM from 'react-dom';
import { cleanup, fireEvent, render } from '@testing-library/react';
import CustomDropzone from './CustomDropzone';
import { useDropzone } from 'react-dropzone';
import { renderHook } from '@testing-library/react-hooks';
import {
  flushPromises,
  createDataTransferWithFiles,
  createFile,
} from './testUtils';

describe('Custom Dropzone Component Test', () => {
  let wrapper;
  let files;

  beforeEach(() => {
    global.URL.createObjectURL = jest.fn();

    files = [createFile('file.pdf', 1024, 'application/pdf')];
    wrapper = render(<CustomDropzone />);
  });

  afterEach(cleanup);

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders', () => {
    expect(wrapper).not.toBeNull();
    // wrapper.debug();
  });

  it('renders without crashing', () => {
    const container = document.createElement('div');
    ReactDOM.render(<CustomDropzone />, container);
    ReactDOM.unmountComponentAtNode(container);
  });

  it('sets {accept} prop on the <input>', () => {
    const accept = 'image/jpeg';
    const { container, debug } = render(<CustomDropzone accept={accept} />);

    const input = container.querySelector('input');

    expect(input).toHaveAttribute('accept', accept);
    // debug();
  });

  it('updates {accept} prop on the <input> when it changes', () => {
    const { container, rerender, debug } = render(
      <CustomDropzone accept="image/jpeg" />
    );

    expect(container.querySelector('input')).toHaveAttribute(
      'accept',
      'image/jpeg'
    );
    // debug();

    rerender(<CustomDropzone accept="image/png" />);

    expect(container.querySelector('input')).toHaveAttribute(
      'accept',
      'image/png'
    );
    // debug();
  });

  it('sets {multiple} prop on the <input>', () => {
    const input = wrapper.container.querySelector('input');
    expect(input).toHaveAttribute('multiple');
    // wrapper.debug();
  });

  it('sets any props passed to the input props getter on the <input>', () => {
    const name = 'dropzone-input';
    const { container, debug } = render(
      <CustomDropzone inputProps={{ name }} />
    );

    const input = container.querySelector('input');
    expect(input).toHaveAttribute('name', name);

    // debug();
  });

  it('sets any props passed to the root props getter on the root node', () => {
    const ariaLabel = 'Dropzone area';
    const { container, debug } = render(
      <CustomDropzone containerProps={{ 'aria-label': ariaLabel }} />
    );

    const dropzone = container.querySelector('div');

    expect(dropzone).toHaveAttribute('aria-label', ariaLabel);

    debug();
  });

  it('runs the custom callback handlers provided to the root props getter', async () => {
    const event = createDataTransferWithFiles(files);

    const rootProps = {
      onClick: jest.fn(),
      onDrop: jest.fn(),
    };

    const ui = <CustomDropzone containerProps={rootProps} />;

    const { container, rerender } = render(ui);

    const dropzone = container.querySelector('div');

    fireEvent.click(dropzone);
    expect(rootProps.onClick).toHaveBeenCalled();

    //TODO: window.GlobalUtilities problem here.
    // fireEvent.drop(dropzone, event);
    // await flushPromises(rerender, ui);
    // expect(rootProps.onDrop).toHaveBeenCalled();
  });

  it('runs the custom callback handlers provided to the input props getter', async () => {
    const inputProps = {
      onClick: jest.fn(),
      onChange: jest.fn(),
    };

    const ui = <CustomDropzone inputProps={inputProps} />;

    const { container, rerender } = render(ui);

    const input = container.querySelector('input');

    fireEvent.click(input);
    await flushPromises(rerender, ui);
    expect(inputProps.onClick).toHaveBeenCalled();

    fireEvent.change(input);
    await flushPromises(rerender, ui);
    expect(inputProps.onChange).toHaveBeenCalled();
  });

  it('runs no callback handlers if {disabled} is true', () => {
    const rootProps = {
      onClick: jest.fn(),
      onKeyDown: jest.fn(),
      onFocus: jest.fn(),
      onBlur: jest.fn(),
    };

    const inputProps = {
      onClick: jest.fn(),
      onChange: jest.fn(),
    };

    const { container, debug } = render(
      <CustomDropzone
        disabled
        inputProps={inputProps}
        containerProps={rootProps}
      />
    );

    const dropzone = container.querySelector('div');

    fireEvent.click(dropzone);
    expect(rootProps.onClick).not.toHaveBeenCalled();

    fireEvent.focus(dropzone);
    fireEvent.keyDown(dropzone);
    expect(rootProps.onFocus).not.toHaveBeenCalled();
    expect(rootProps.onKeyDown).not.toHaveBeenCalled();

    fireEvent.blur(dropzone);
    expect(rootProps.onBlur).not.toHaveBeenCalled();

    const input = container.querySelector('input');

    fireEvent.click(input);
    expect(inputProps.onClick).not.toHaveBeenCalled();

    fireEvent.change(input);
    expect(inputProps.onChange).not.toHaveBeenCalled();

    // debug();
  });

  test('{rootRef, inputRef} are exposed', () => {
    const { result } = renderHook(() => useDropzone());
    const { rootRef, inputRef, getRootProps, getInputProps } = result.current;

    const { container } = render(
      <div {...getRootProps()}>
        <input {...getInputProps()} />
      </div>
    );

    expect(container.querySelector('div')).toEqual(rootRef.current);
    expect(container.querySelector('input')).toEqual(inputRef.current);
  });

  test('{tabindex} is 0 if {disabled} is false', () => {
    const { container, debug } = render(<CustomDropzone />);
    expect(container.querySelector('div')).toHaveAttribute('tabindex', '0');

    // debug();
  });

  test('{tabindex} is not set if {disabled} is true', () => {
    const { container, rerender, debug } = render(<CustomDropzone />);

    expect(container.querySelector('div')).toHaveAttribute('tabindex', '0');

    // debug();

    rerender(<CustomDropzone disabled />);

    expect(container.querySelector('div')).not.toHaveAttribute('tabindex');

    // debug();
  });
});
