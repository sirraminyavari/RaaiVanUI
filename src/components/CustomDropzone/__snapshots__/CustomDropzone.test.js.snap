// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Dropzone Component Test matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ hrLASB"
        tabindex="0"
      >
        <div
          class="sc-gsTCUz dbQMNh"
        >
          <svg
            color="#fff"
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(255, 255, 255);"
            viewBox="0 0 24 24"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 12.586l4.243 4.242-1.415 1.415L13 16.415V22h-2v-5.587l-1.828 1.83-1.415-1.415L12 12.586zM12 2a7.001 7.001 0 0 1 6.954 6.194 5.5 5.5 0 0 1-.953 10.784L18 17a6 6 0 0 0-11.996-.225L6 17v1.978a5.5 5.5 0 0 1-.954-10.784A7 7 0 0 1 12 2z"
              />
            </g>
          </svg>
        </div>
        <div
          class="sc-dlfnbm eKRbBI"
        >
          <input
            autocomplete="off"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <p>
            برای انتخاب فایل آنها را بگیرید و درون باکس رها و یا کلیک کنید...
          </p>
        </div>
      </div>
      <aside
        class="sc-hKgILt kPbuTm"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ hrLASB"
      tabindex="0"
    >
      <div
        class="sc-gsTCUz dbQMNh"
      >
        <svg
          color="#fff"
          fill="currentColor"
          height="30"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 12.586l4.243 4.242-1.415 1.415L13 16.415V22h-2v-5.587l-1.828 1.83-1.415-1.415L12 12.586zM12 2a7.001 7.001 0 0 1 6.954 6.194 5.5 5.5 0 0 1-.953 10.784L18 17a6 6 0 0 0-11.996-.225L6 17v1.978a5.5 5.5 0 0 1-.954-10.784A7 7 0 0 1 12 2z"
            />
          </g>
        </svg>
      </div>
      <div
        class="sc-dlfnbm eKRbBI"
      >
        <input
          autocomplete="off"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        <p>
          برای انتخاب فایل آنها را بگیرید و درون باکس رها و یا کلیک کنید...
        </p>
      </div>
    </div>
    <aside
      class="sc-hKgILt kPbuTm"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
